service/history/workflow/cache/cache.go

```diff
@@ -198,16 +202,39 @@ func (c *CacheImpl) getOrCreateWorkflowExecutionInternal(
 	//  Consider revisiting this if it causes too much GC activity
 	releaseFunc := c.makeReleaseFunc(key, workflowCtx, forceClearContext, lockPriority)
 
-	if err := workflowCtx.Lock(ctx, lockPriority); err != nil {
-		// ctx is done before lock can be acquired
-		c.Release(key)
+	if err := c.lockWorkflowExecution(ctx, workflowCtx, key, lockPriority); err != nil {
 		handler.Counter(metrics.CacheFailures.GetMetricName()).Record(1)
 		handler.Counter(metrics.AcquireLockFailedCounter.GetMetricName()).Record(1)
 		return nil, nil, err
 	}
+
 	return workflowCtx, releaseFunc, nil
 }
 
+func (c *CacheImpl) lockWorkflowExecution(ctx context.Context,
+	workflowCtx workflow.Context,
+	key definition.WorkflowKey,
+	lockPriority workflow.LockPriority) error {
+
+	var cancel context.CancelFunc
+	if headers.GetCallerInfo(ctx).CallerType != headers.CallerTypeAPI {
+		timeout := time.Now().Add(nonApiContextLockTimeout)
+		ctx, cancel = context.WithDeadline(ctx, timeout)
+		defer cancel()
+	} else if deadline, ok := ctx.Deadline(); ok {
+		timeout := deadline.Add(-workflowLockTimeoutTailTime)
+		ctx, cancel = context.WithDeadline(ctx, timeout)
+		defer cancel()
+	}
+
+	if err := workflowCtx.Lock(ctx, lockPriority); err != nil {
+		// ctx is done before lock can be acquired
+		c.Release(key)
+		return serviceerror.NewResourceExhausted(enums.RESOURCE_EXHAUSTED_CAUSE_BUSY_WORKFLOW, "Workflow cannot be locked")
```

Comment:
We need to mark this resource exhausted cause as non-retryable for rpc client & handler in server I think, especially for background task/call, otherwise the rpc call will still hang there try to get lock & backoff & retry and block the caller goroutine.

https://github.com/temporalio/temporal/blob/master/common/util.go#L334

Reply:
Done